pkg/xds/server/callbacks/dataplane_sync_callbacks.go

```diff
@@ -0,0 +1,230 @@
+	if request.NodeId() == "" {
+		// from https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack-nack-and-versioning:
+		// Only the first request on a stream is guaranteed to carry the node identifier.
+		// The subsequent discovery requests on the same stream may carry an empty node identifier.
+		// This holds true regardless of the acceptance of the discovery responses on the same stream.
+		// The node identifier should always be identical if present more than once on the stream.
+		// It is sufficient to only check the first message for the node identifier as a result.
+		return nil
+	}
+
+	t.RLock()
+	dataplaneSyncLog.V(1).Info("stream request", "streamID", streamID)
+	alreadyProcessed := t.activeStreams[streamID].proxyInfo != nil
+	t.RUnlock()
+	if alreadyProcessed { // We fast return if we already know that this streamID is tracking a specific dataplane
+		return nil
+	}
+
+	proxyId, err := core_xds.ParseProxyIdFromString(request.NodeId())
+	if err != nil {
+		return errors.Wrap(err, "invalid node ID")
+	}
+	dpKey := proxyId.ToResourceKey()
+	metadata := core_xds.DataplaneMetadataFromXdsMetadata(request.Metadata())
+	if metadata == nil {
+		return errors.New("metadata in xDS Node cannot be nil")
+	}
+	if metadata.Resource != nil {
+		if core_model.MetaToResourceKey(metadata.Resource.GetMeta()) != dpKey {
+			return fmt.Errorf("proxyId %s does not match proxy resource %s", dpKey, metadata.Resource.GetMeta())
+		}
+	}
+	l := dataplaneSyncLog.WithValues("dpKey", dpKey, "streamID", streamID)
+	t.Lock()
+	if t.activeStreams[streamID].proxyInfo != nil {
+		return nil // We fast return if we already know that this streamID is tracking a specific dataplane
+	}
+	pInfo := t.proxyInfos[dpKey]
+	if pInfo == nil {
+		ctx, cancel := context.WithCancel(context.Background())
+		done := make(chan struct{})
+		pInfo = &dpInfo{
+			dpKey:            dpKey,
+			lifecycleManager: t.dpLifecycleManagerFactory.New(dpKey),
+			ctx:              ctx,
+			done:             done,
+			meta:             atomic.Pointer[core_xds.DataplaneMetadata]{},
+			cancelFunc:       cancel,
+		}
+		t.proxyInfos[dpKey] = pInfo
+	}
+	t.activeStreams[streamID].proxyInfo = pInfo
+	ctx := t.activeStreams[streamID].ctx
+	pInfo.Lock()
+	defer pInfo.Unlock()
+	t.Unlock()
+	pInfo.streams = append(pInfo.streams, streamID)
+	pInfo.meta.Store(metadata)
+	if len(pInfo.streams) == 1 {
+		err := pInfo.lifecycleManager.Register(l, ctx, metadata)
+		if err != nil {
+			return err
+		}
+
+		l.V(1).Info("starting watchdog")
+		go func() {
+			defer close(pInfo.done)
+			l.V(1).Info("watchdog started")
+			if t.dpWatchdogFactory != nil {
+				t.dpWatchdogFactory.New(dpKey, pInfo.meta.Load).Start(pInfo.ctx)
+			}
+		}()
+	} else {
+		l.V(1).Info("watchdog was started by a previous stream, not starting a new one", "activeStreams", pInfo.streams)
+	}
+
+	return nil
+}
+
+func (t *dataplaneSyncCallbacks) OnStreamClosed(streamID core_xds.StreamID) {
+	t.Lock()
```

Comment:
So many lock/unlock invocations in this method. I'm thinking how can we improve it to make it cleaner.